### PR TITLE
PIM-6647: Define 2 separate indexes for P and P&PM

### PR DESCRIPTION
**Pull request by SamirBoulil**
**Description (for Contributor and Core Developer)**

This PR adds the support for a second index in the PIM that will be used to index products and product models.

- A new way to define a index configuration (config.yml)
- Add a way to instanciate dynamically ES clients based on those configurations
- Adapts the way the indexed are reset in the pim:install and in our test stack.

The pim installs and creates both of those indexes.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Ok
| Added Behats                      | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Bundle/ElasticsearchBundle/ClientInterface.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
+<?php
```
Comment (SamirBoulil) [inline src/Akeneo/Bundle/ElasticsearchBundle/ClientInterface.php]: To remove

**Comment by SamirBoulil** [comment]
Review with JJ:
- Remove ClientInterface for now
- Move the ProductClientRegistry from ESBundle to CatalogBundle (as it's tinted Business and not technical)
- rename the index configuration for a list of (index_name, index_conf)

File: src/Akeneo/Bundle/ElasticsearchBundle/DependencyInjection/AkeneoElasticsearchExtension.php [hunk 2]
```diff
@@ -24,11 +27,24 @@ public function load(array $configs, ContainerBuilder $container)
...
+                'akeneo_elasticsearch.index_configuration.%s.files',
+                $index['index_name']
+            );
+            $container->register($configurationLoaderServiceName, Loader::class)
+                ->setArguments([$index['configuration_files']]);
 
-        $loader = new Loader\YamlFileLoader($container, new FileLocator(__DIR__.'/../Resources/config'));
+            $container->register($index['service_name'], Client::class)
+                ->setArguments([
+                    new Reference('akeneo_elasticsearch.client_builder'), // $clientBuilder
```
Comment (SamirBoulil) [inline src/Akeneo/Bundle/ElasticsearchBundle/DependencyInjection/AkeneoElasticsearchExtension.php]: to remove //$clientBuilder

File: src/Akeneo/Bundle/ElasticsearchBundle/DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -24,26 +24,36 @@ public function getConfigTreeBuilder()
...
-                ->end()
                 ->arrayNode('hosts')
                     ->prototype('scalar')->defaultValue(['localhost:9200'])->end()
                     ->isRequired()->cannotBeEmpty()
                     ->beforeNormalization()
-                        ->ifString()
-                        ->then(function ($value) {
-                            return [$value];
-                        })
+                    ->ifString()
```
Comment (SamirBoulil) [inline src/Akeneo/Bundle/ElasticsearchBundle/DependencyInjection/Configuration.php]: indent

File: src/Akeneo/Bundle/ElasticsearchBundle/DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -24,26 +24,36 @@ public function getConfigTreeBuilder()
...
+                            ->end()
+                            ->arrayNode('configuration_files')
+                                ->isRequired()
+                                ->requiresAtLeastOneElement()
+                                ->prototype('scalar')->isRequired()->cannotBeEmpty()->end()
+                                ->info('Paths of the YAML files to configure the index. See https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-create-index.html and src/Akeneo/Bundle/ElasticsearchBundle/IndexConfiguration/IndexConfiguration.php.')
+                            ->end()
+                            ->scalarNode('service_name')
+                                ->isRequired()->cannotBeEmpty()
+                                ->info('Name of the symfony service for this client.')
```
Comment (SamirBoulil) [inline src/Akeneo/Bundle/ElasticsearchBundle/DependencyInjection/Configuration.php]: Say that the service is automatically register in the container + update README.md ElasticSearchBundle

File: src/Pim/Bundle/CatalogBundle/Resources/config/cursors.yml [hunk 5]
```diff
@@ -11,7 +11,17 @@ services:
...
         arguments:
-            - '@akeneo_elasticsearch.client'
+            - '@akeneo_elasticsearch.client.product'
+            - '@pim_catalog.object_manager.product'
+            - '%pim_catalog.entity.product.class%'
+            - '%akeneo_elasticsearch.cursor.cursor.class%'
+            - '%pim_catalog.factory.product_cursor.page_size%'
+            - 'pim_catalog_product'
+
+    pim_catalog.factory.product_and_model_cursor:
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Resources/config/cursors.yml]: To remove: no need

File: src/Pim/Bundle/CatalogBundle/Resources/config/cursors.yml [hunk 6]
```diff
@@ -31,7 +51,17 @@ services:
...
         arguments:
-            - '@akeneo_elasticsearch.client'
+            - '@akeneo_elasticsearch.client.product'
+            - '@pim_catalog.object_manager.product'
+            - '%pim_catalog.entity.product.class%'
+            - '%akeneo_elasticsearch.cursor.from_size_cursor.class%'
+            - '%pim_catalog.factory.product_cursor.page_size%'
+            - 'pim_catalog_product'
+
+    pim_catalog.factory.product_and_model_from_size_cursor:
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Resources/config/cursors.yml]: to remove

File: src/Pim/Bundle/CatalogBundle/Resources/config/cursors.yml [hunk 7]
```diff
@@ -11,7 +11,17 @@ services:
     pim_catalog.factory.product_cursor:
         class: '%akeneo_elasticsearch.cursor.cursor_factory.class%'
         arguments:
-            - '@akeneo_elasticsearch.client'
+            - '@akeneo_elasticsearch.client.product'
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Resources/config/cursors.yml]: Revert and rename with right service

Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Resources/config/cursors.yml]: done

File: CHANGELOG-1.8.md [hunk 8]
```diff
@@ -9,6 +9,7 @@
 
 ## Technical improvements
 
+- PIM-6647: Define two separate indexes for products and products and product models
```
Comment (damien-carcel) [inline CHANGELOG-1.8.md]: Is it usefull to add that in the Changelog, as even having one is a new feature anyway?

Comment (SamirBoulil) [inline CHANGELOG-1.8.md]: Yes i'll revert it ! 👍

**Pull request merged**